<a href="https://colab.research.google.com/github/treyra/CS155_miniproject/blob/master/Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# This notebook creates a simple logistic regression model on the data to experiment with different features and different normilizations

In [19]:
import numpy as np
from matplotlib import pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold
import pandas as pd

### Load the data:

In [12]:
#Use the imputed and normalized data (This link will need to be updated when we merge the branches)
dfImp = pd.read_csv('https://raw.githubusercontent.com/treyra/CS155_miniproject/data_visualization_v1/data_cleanup/traindata_imputed_normalized_minmax.csv',sep=',', header=0)
dfImp.head()

Unnamed: 0  id  LATITUDE  LONGITUDE  STATE  ...  LABEL    0    1    2    3
0           0   0  0.666803  -2.100242      1  ...      1  1.0  0.0  0.0  0.0
1           1   1  0.590150  -1.485350      2  ...      4  0.0  0.0  0.0  1.0
2           2   2  0.562013  -1.439874      2  ...      2  0.0  1.0  0.0  0.0
3           3   3  0.567394  -1.426084      2  ...      4  0.0  0.0  0.0  1.0
4           4   4  0.587546  -2.027583      1  ...      2  0.0  1.0  0.0  0.0

[5 rows x 17 columns]

###Create the catgegory vs all data

In [16]:
#Make a copy for each label, set that labels value to 1, the rest to -1
#To change the features being passed in, make a dataframe with the new features, and replace dfImp with it.

oneData = dfImp.copy()
oneData.loc[(oneData['LABEL'] != 1,"LABEL")] = -1
#Make numpy arrays for regression
oneY = oneData['LABEL'].to_numpy()
oneData.drop('LABEL', axis=1)
oneX = oneData.to_numpy()

twoData = dfImp.copy()
twoData.loc[(twoData['LABEL'] != 2,"LABEL")] = -1
twoData.loc[(twoData['LABEL'] == 2,"LABEL")] = 1
#Make numpy arrays for regression
twoY = twoData['LABEL'].to_numpy()
twoData.drop('LABEL', axis=1)
twoX = twoData.to_numpy()

threeData = dfImp.copy()
threeData.loc[(threeData['LABEL'] != 3,"LABEL")] = -1
threeData.loc[(threeData['LABEL'] == 3,"LABEL")] = 1
threeY = threeData['LABEL'].to_numpy()
threeData.drop('LABEL', axis=1)
threeX = threeData.to_numpy()

fourData = dfImp.copy()
fourData.loc[(fourData['LABEL'] != 4,"LABEL")] = -1
fourData.loc[(fourData['LABEL'] == 4,"LABEL")] = 1
fourY = fourData['LABEL'].to_numpy()
fourData.drop('LABEL', axis=1)
fourX = fourData.to_numpy()

In [18]:
len(oneX)

285382

###Logistic Regression and Cross Validation

In [ ]:
#Regularization values to check via cross validation
#This can be swapped with some other parameter, the approach will be the came
numParams = 20
Cs = np.logspace(-1,1,numParams)

#Scores for in and out classification for each one vs all, determined using LogisticRegression.score()
scoreIn = np.zeros((4,numParams))
scoreOut = np.zeros((4,numParams))

#Run cross validation for each parameter!
num_folds = 10
for run in range(numParams):

    # Initialize kfold cross-validation object with 10 folds:
    kf = KFold(n_splits=num_folds)

    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(oneX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = oneX[train_index], oneX[test_index]
        y_train, y_test = oneY[train_index], oneY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[0,run] += log.score(x_train, y_train)
        scoreOut[0,run] += log.score(x_train, y_train)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(twoX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = twoX[train_index], twoX[test_index]
        y_train, y_test = twoY[train_index], twoY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[1,run] += log.score(x_train, y_train)
        scoreOut[1,run] += log.score(x_train, y_train)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(threeX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = threeX[train_index], threeX[test_index]
        y_train, y_test = threeY[train_index], threeY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[2,run] += log.score(x_train, y_train)
        scoreOut[2,run] += log.score(x_train, y_train)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(fourX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = fourX[train_index], fourX[test_index]
        y_train, y_test = fourY[train_index], fourY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[3,run] += log.score(x_train, y_train)
        scoreOut[3,run] += log.score(x_train, y_train)

scoreIn = scoreIn/num_folds
scoreOut = scoreOut/num_folds

print("Scores In")
print(scoreIn)
print("Scores Out")
print(scoreOut)



In [23]:


scoreIn = scoreIn/num_folds
scoreOut = scoreOut/num_folds

print("Scores In")
print(scoreIn)
print("Scores Out")
print(scoreOut)

Scores In
[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294]
 [0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896 ]
 [0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826]
 [0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867]]
Scores Out
[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.912922

Output from first run when Cs was accidentally set to .1-10**10

Scores In

[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294]

 [0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896 ]

 [0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826]
 [0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867]]

Scores Out

[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294]

 [0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896 ]

 [0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826]

 [0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867]]

In [27]:
Cs,np.logspace(-1,1,numParams)

(array([1.25892541e+00, 4.17881006e+00, 1.38709199e+01, 4.60423938e+01,
        1.52830673e+02, 5.07298009e+02, 1.68389803e+03, 5.58944158e+03,
        1.85532951e+04, 6.15848211e+04, 2.04421380e+05, 6.78545457e+05,
        2.25232770e+06, 7.47625680e+06, 2.48162892e+07, 8.23738707e+07,
        2.73427446e+08, 9.07600522e+08, 3.01264090e+09, 1.00000000e+10]),
 array([ 0.1       ,  0.1274275 ,  0.16237767,  0.20691381,  0.26366509,
         0.33598183,  0.42813324,  0.54555948,  0.6951928 ,  0.88586679,
         1.12883789,  1.43844989,  1.83298071,  2.33572147,  2.97635144,
         3.79269019,  4.83293024,  6.15848211,  7.8475997 , 10.        ]))